syft/cataloger/python/package_cataloger.go

```diff
@@ -32,33 +38,87 @@ func (c *PackageCataloger) Name() string {
+	for _, entry := range entries {
+		p, err := c.catalogEggOrWheel(entry)
 		if err != nil {
-			return nil, fmt.Errorf("failed to find files by glob: %s", glob)
+			return nil, fmt.Errorf("unable to catalog python package=%+v: %w", entry.Metadata.Location.Path, err)
+		}
+		if p != nil {
+			packages = append(packages, *p)
 		}
-		fileMatches = append(fileMatches, matches...)
 	}
 
-	var pkgs []pkg.Package
-	for _, location := range fileMatches {
-		p, err := c.catalogEggOrWheel(resolver, location)
-		if err != nil {
-			return nil, fmt.Errorf("unable to catalog python package=%+v: %w", location.Path, err)
+	return packages, nil
+}
+
+func (c *PackageCataloger) getPythonPackageEntries(resolver source.Resolver) ([]*pythonPackageData, error) {
+	var metadataLocations []source.Location
+
+	// find all primary record paths
+	matches, err := resolver.FilesByGlob(eggMetadataGlob, wheelMetadataGlob)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find files by glob: %w", err)
+	}
+	metadataLocations = append(metadataLocations, matches...)
+
+	// for every primary record path, craft all secondary record paths and build a request object to gather all file contents for each record
+	var requester = source.NewContentRequester()
+	var entries = make([]*pythonPackageData, len(metadataLocations))
+	for i, metadataLocation := range metadataLocations {
+		// we've been given a file reference to a specific wheel METADATA file. note: this may be for a directory
+		// or for an image... for an image the METADATA file may be present within multiple layers, so it is important
+		// to reconcile the RECORD path to the same layer (or the next adjacent lower layer). The same is true with
+		// the top_level.txt file.
+
+		// lets find the RECORD file relative to the directory where the METADATA file resides (in path AND layer structure)
+		recordPath := filepath.Join(filepath.Dir(metadataLocation.Path), "RECORD")
+		recordLocation := resolver.RelativeFileByPath(metadataLocation, recordPath)
+
+		// a top_level.txt file specifies the python top-level packages (provided by this python package) installed into site-packages
+		parentDir := filepath.Dir(metadataLocation.Path)
+		topLevelPath := filepath.Join(parentDir, "top_level.txt")
+		topLevelLocation := resolver.RelativeFileByPath(metadataLocation, topLevelPath)
+
+		// build an entry that will later be populated with contents when the request is executed
+		entry := &pythonPackageData{
+			Metadata: source.FileData{
+				Location: metadataLocation,
+			},
 		}
-		if p != nil {
-			pkgs = append(pkgs, *p)
+
+		requester.Add(&entry.Metadata)
+
+		if recordLocation != nil {
+			entry.FileRecord = &source.FileData{
+				Location: *recordLocation,
+			}
+			requester.Add(entry.FileRecord)
+		}
+
+		if topLevelLocation != nil {
+			entry.TopPackage = &source.FileData{
+				Location: *topLevelLocation,
+			}
+			requester.Add(entry.TopPackage)
 		}
+
+		// keep the entry for processing later
+		entries[i] = entry
```

Comment:
This logic in the body of the `for` loop could potentially be factored out into a dedicated function

Reply:
yes, I can make a helper function called `getPythonPackageEntry` which would improve readability a bit